File: Day_51/LowestCommonAncestorBST.c
#include <stdio.h>
#include <stdlib.h>

// Definition for a BST node
struct Node {
    int data;
    struct Node *left;
    struct Node *right;
};

// Helper function to create a new node
struct Node* createNode(int value) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = value;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

// Function to insert a value into the BST
struct Node* insert(struct Node* root, int value) {
    if (root == NULL) {
        return createNode(value);
    }
    if (value < root->data) {
        root->left = insert(root->left, value);
    } else if (value > root->data) {
        root->right = insert(root->right, value);
    }
    return root;
}

// Function to find the Lowest Common Ancestor (LCA) in a BST
struct Node* findLCA(struct Node* root, int n1, int n2) {
    if (root == NULL) {
        return NULL;
    }

    // If both n1 and n2 are strictly smaller than root, LCA must be in the left subtree
    if (root->data > n1 && root->data > n2) {
        return findLCA(root->left, n1, n2);
    }
    
    // If both n1 and n2 are strictly greater than root, LCA must be in the right subtree
    if (root->data < n1 && root->data < n2) {
        return findLCA(root->right, n1, n2);
    }
    
    // If one is smaller and one is greater (or one is equal to root), 
    // the current root is the Lowest Common Ancestor.
    return root;
}

int main() {
    int n;
    
    // Read the number of elements
    if (scanf("%d", &n) != 1) {
        return 0;
    }

    struct Node* root = NULL;
    int val;

    // Read the elements and insert them into the BST
    for (int i = 0; i < n; i++) {
        scanf("%d", &val);
        root = insert(root, val);
    }

    int n1, n2;
    // Read the two target nodes
    scanf("%d %d", &n1, &n2);

    // Find and print the LCA
    struct Node* lca = findLCA(root, n1, n2);
    
    if (lca != NULL) {
        printf("%d\n", lca->data);
    } else {
        printf("LCA not found\n");
    }

    return 0;
}

File: Day_51/LowestCommonAncestorBST.java
public class LowestCommonAncestorBST {

    // Definition for a binary tree node
    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) { 
            val = x; 
        }
    }

    // Function to find the Lowest Common Ancestor
    public static TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        // If both p and q are smaller than root, LCA must be in the left subtree
        if (p.val < root.val && q.val < root.val) {
            return lowestCommonAncestor(root.left, p, q);
        }
        
        // If both p and q are greater than root, LCA must be in the right subtree
        if (p.val > root.val && q.val > root.val) {
            return lowestCommonAncestor(root.right, p, q);
        }
        
        // If they split, or one equals the root, the current root is the LCA
        return root;
    }

    public static void main(String[] args) {
        // Creating the BST from the classic LeetCode example:
        //
        //          6
        //        /   \
        //       2     8
        //      / \   / \
        //     0   4 7   9
        //        / \
        //       3   5

        TreeNode root = new TreeNode(6);
        root.left = new TreeNode(2);
        root.right = new TreeNode(8);
        
        root.left.left = new TreeNode(0);
        root.left.right = new TreeNode(4);
        
        root.right.left = new TreeNode(7);
        root.right.right = new TreeNode(9);
        
        root.left.right.left = new TreeNode(3);
        root.left.right.right = new TreeNode(5);

        // Test Case 1: p = 2, q = 8
        TreeNode p1 = root.left;  // Node 2
        TreeNode q1 = root.right; // Node 8
        TreeNode lca1 = lowestCommonAncestor(root, p1, q1);
        System.out.println("LCA of " + p1.val + " and " + q1.val + " is: " + lca1.val);

        // Test Case 2: p = 2, q = 4
        TreeNode p2 = root.left;        // Node 2
        TreeNode q2 = root.left.right;  // Node 4
        TreeNode lca2 = lowestCommonAncestor(root, p2, q2);
        System.out.println("LCA of " + p2.val + " and " + q2.val + " is: " + lca2.val);
    }
}